flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/FlinkJobReconciler.java

```diff
@@ -0,0 +1,213 @@
+            FlinkConfigManager configManager,
+            EventRecorder eventRecorder) {
+        super(kubernetesClient, flinkService, configManager, eventRecorder);
+    }
+
+    @Override
+    public boolean reconcileInternal(CR resource, Context context) throws Exception {
+        var deployMeta = resource.getMetadata();
+        STATUS status = resource.getStatus();
+        var reconciliationStatus = status.getReconciliationStatus();
+        SPEC lastReconciledSpec = reconciliationStatus.deserializeLastReconciledSpec();
+        SPEC currentDeploySpec = resource.getSpec();
+
+        var deployConfig = getDeployConfig(deployMeta, currentDeploySpec, context);
+        if (shouldWaitForPendingSavepoint(status.getJobStatus(), deployConfig)) {
+            LOG.info("Delaying job reconciliation until pending savepoint is completed.");
+            return true;
+        }
+
+        var observeConfig = getObserveConfig(resource, context);
+        boolean specChanged = !currentDeploySpec.equals(lastReconciledSpec);
+        if (specChanged) {
+            if (checkNewSpecAlreadyDeployed(resource, deployConfig)) {
+                return true;
+            }
+            LOG.debug("Detected spec change, starting upgrade process.");
+            JobState currentJobState = lastReconciledSpec.getJob().getState();
+            JobState desiredJobState = currentDeploySpec.getJob().getState();
+            JobState newState = currentJobState;
+            if (currentJobState == JobState.RUNNING) {
+                if (desiredJobState == JobState.RUNNING) {
+                    LOG.info("Upgrading/Restarting running job, suspending first...");
+                }
+                Optional<UpgradeMode> availableUpgradeMode =
+                        getAvailableUpgradeMode(resource, deployConfig, observeConfig);
+                if (availableUpgradeMode.isEmpty()) {
+                    return true;
+                }
+                // We must record the upgrade mode used to the status later
+                currentDeploySpec.getJob().setUpgradeMode(availableUpgradeMode.get());
+                cancelJob(resource, availableUpgradeMode.get(), observeConfig);
+                newState = JobState.SUSPENDED;
+            }
+            if (currentJobState == JobState.SUSPENDED && desiredJobState == JobState.RUNNING) {
+                restoreJob(
+                        deployMeta,
+                        currentDeploySpec,
+                        status,
+                        deployConfig,
+                        // We decide to enforce HA based on how job was previously suspended
+                        lastReconciledSpec.getJob().getUpgradeMode() == UpgradeMode.LAST_STATE);
+                newState = JobState.RUNNING;
+            }
+            ReconciliationUtils.updateForSpecReconciliationSuccess(
+                    resource, newState, deployConfig);
+        } else if (shouldRollBack(reconciliationStatus, observeConfig)) {
```

Comment:
Is `shouldRollback` and `rollbackApplication` need to be abstract method to let the Application and sessionJob to implement ?